File: cpp/2670_distinct_difference_array/distinct_difference_array.cpp
/*
 * @Date: 2024-01-31
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-01-31
 * @FilePath: /algorithm/cpp/2670_distinct_difference_array/distinct_difference_array.cpp
 */

#include <cassert>
#include <tuple>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> distinctDifferenceArray(vector<int>& nums) {
    int n = nums.size();
    unordered_set<int> st;
    vector<int> sufCnt(n + 1, 0);
    for (int i = n - 1; i > 0; i--) {
      st.insert(nums[i]);
      sufCnt[i] = st.size();
    }
    vector<int> res;
    st.clear();
    for (int i = 0; i < n; i++) {
      st.insert(nums[i]);
      res.push_back(int(st.size()) - sufCnt[i + 1]);
    }
    return res;
  }
};

int main() {
  vector<tuple<vector<int>, vector<int>>> tests{
      {{1, 2, 3, 4, 5}, {-3, -1, 1, 3, 5}},
      {{3, 2, 3, 4, 2}, {-2, -1, 0, 2, 3}},
  };

  for (auto& [nums, ans] : tests) {
    assert(Solution().distinctDifferenceArray(nums) == ans);
  }
}

File: rust/2670_distinct_difference_array/distinct_difference_array.rs
/*
 * @Date: 2024-01-31
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-01-31
 * @FilePath: /algorithm/rust/2670_distinct_difference_array/distinct_difference_array.rs
 */

struct Solution;

impl Solution {
    pub fn distinct_difference_array(nums: Vec<i32>) -> Vec<i32> {
        use std::collections::HashSet;
        let n = nums.len();
        let mut suf = vec![0; n + 1];
        let mut s = HashSet::new();

        for i in (0..n).rev() {
            s.insert(nums[i]);
            suf[i] = s.len();
        }

        let mut ans = Vec::new();
        s.clear();
        for i in 0..n {
            s.insert(nums[i]);
            ans.push((s.len() - suf[i + 1]) as i32);
        }

        ans
    }
}

fn main() {
    let tests = vec![
        (vec![1, 2, 3, 4, 5], vec![-3, -1, 1, 3, 5]),
        (vec![3, 2, 3, 4, 2], vec![-2, -1, 0, 2, 3]),
    ];

    for (nums, ans) in tests {
        assert_eq!(Solution::distinct_difference_array(nums), ans);
    }
}
